debugging.md

```diff
@@ -0,0 +1,107 @@
+# Debugging
+
+This document aims to collect info on the current status and future opportunities for debugging Rust.
+
+## Debugging architecture overview
+
+The compiler generates debuginfo for Rust programs (see https://github.com/rust-lang/rust/tree/master/src/librustc_trans/debuginfo).
+This is an LLVM-specific thing. LLVM then processes this through optimisations
+(with varying degrees of success) and outputs debug symbols with the library.
+These are DWARF for the Linux, Mac, and Windows/GNU toolchains, and pdb (program
+database) for Windows/visual studio (pdb support is WIP in LLVM, I believe).
+
+Debuggers consume the generated symbols. Visual Studio only works on Windows.
+LLDB works everywhere but is primarily used as a backend for XCode on MacOS. GDB
+works everywhere. GDB and LLDB are command
+line tools, but can have frontends.
+
+Debugger support can come in various forms in this architecture:
+* we could invent our own debugger and debug symbols (this is infeasible due to resourcing)
+* generate better debuginfo in the compiler - most of the low hanging fruit here is done
+* extend debug symbols with Rust-specific forms
+* improve the debuggers handling of Rust
+  - plugins (I think this is primarily an LLDB thing)
+  - land code in the debuggers (done to some extent in GDB)
+  - improve the debuggers *and* the debug symbols
+  - pretty printers (used by debuggers to display rust data, see gdb and lldb pretty printers in https://github.com/rust-lang/rust/tree/master/src/etc)
+
+
+
+## Pretty printers
+
+* need to run via the scripts (in https://github.com/rust-lang/rust/tree/master/src/etc))
+* still needed even with upstream improvements
+  - not sure if there is conflicts
+* gdb and lldb have quite different APIs
+
+
+## Visual Studio support
+
+* we support as much as llvm, but llvm is incomplete
+* no CI for testing this
+* http://www.jonathanturner.org/2017/03/rust-in-windows.html
+
+
+## Potential improvements
+
+### Debug info
+
+Better support for trait objects and method calls. TODO would this require DWARF extensions?
```

Comment:
FWIW I'm working on printing trait objects.  It's slow going due to my limited time, and kind of a pain to track my progress due to the different places that have to be touched.

Anyway, at least for printing trait objects, only a minor DWARF extension is needed; specifically, a way to link a vtable to the concrete type for which it was created.  That is, no new tag is needed, just the ability to put a DW_AT_containing_type on the vtable object's DIE.

I haven't looked deeply into the other part yet, aside from what we were working on in that google doc about Rust representation in DWARF.